### ISSUE TITLE
Research: Image Generation for Dynamic Expressions

### ISSUE THREAD

**Issue by suparious** [suparious/adult-waifu-animation#4]
## Overview
Research image/video generation to create dynamic character expressions instead of static sprites.

## Current State
- Static 2D images per character (luna.png, sakura.png)
- Three.js mesh deformation for basic animation
- No dynamic expression generation

## Vision
- Generate emotion-specific expressions on demand
- Create outfit variations dynamically
- Animate transitions between expressions
- Eventually: full video generation for complex animations

### Image Generation
| Technology | Use Case | Notes |
|------------|----------|-------|
| **Stable Diffusion** | Base image generation | Self-hostable on GPU |
| **SDXL** | Higher quality | More VRAM needed |
| **ControlNet** | Pose/expression control | Consistent character |
| **IP-Adapter** | Character consistency | Face/style transfer |
| **LoRA** | Character fine-tuning | Train custom waifus |

### Animation
| Technology | Use Case | Notes |
|------------|----------|-------|
| **AnimateDiff** | Image-to-video | Short clips |
| **Deforum** | Morphing animations | Trippy but useful |
| **Live2D** | 2D rigged animation | Industry standard for VTubers |
| **Motion LoRA** | Animated SD | Experimental |

## Research Tasks
- [ ] Test SD 1.5 / SDXL on existing GPU nodes (RTX 3060 12GB)
- [ ] Evaluate ControlNet for expression control
- [ ] Research character consistency techniques
- [ ] Benchmark generation latency
- [ ] Design caching strategy for common expressions
- [ ] Investigate Live2D as alternative approach

## Infrastructure Needs
- GPU allocation for inference
- Model storage (NFS or S3)
- Queue system for generation requests

## Related
- Drives: SolidRusT.AI image generation feature
- Consider: ComfyUI as backend
- Consider: Integration with voice for lip-sync

**Comment by suparious** [comment]
## Research Findings: Image Generation for Dynamic Expressions

I've evaluated image generation technologies for creating dynamic waifu expressions. Here's the analysis:

### Recommended Approach: ComfyUI + ControlNet + IP-Adapter

**Architecture:**
```
Static base images (current) → ControlNet pose guidance → IP-Adapter character consistency → SDXL refinement → Dynamic expressions
```

This provides the best balance of quality, control, and GPU efficiency for the SolidRusT infrastructure.

---

## GPU Capacity Analysis

**Available Resources (SolidRusT HQ):**
- 5x RTX 3060 12GB nodes
- Current allocation:
  - vLLM (Qwen3-4B): 1x GPU
  - vLLM Embeddings (bge-m3): 1x GPU
  - **Available for image gen: 3x GPUs**

**Stable Diffusion Requirements:**
| Model | VRAM | Speed | Quality | Batch Size |
|-------|------|-------|---------|------------|
| SD 1.5 | 4GB | 3-5s | Good | 1-2 images |
| SDXL | 8GB | 8-12s | Excellent | 1 image |
| SDXL Turbo | 8GB | 2-3s | Excellent | 1 image |

**Verdict:** RTX 3060 12GB is perfect for SDXL with ControlNet + IP-Adapter.

---

## Recommended Stack

### 1. ComfyUI (Workflow Engine) - TOP CHOICE

**Why ComfyUI?**
- ✅ **Node-based workflows** - Visual workflow editor
- ✅ **Production-ready API** - JSON workflow execution
- ✅ **All features built-in** - ControlNet, IP-Adapter, LoRA support
- ✅ **Excellent caching** - Model stays loaded in VRAM
- ✅ **Active development** - Weekly updates
- ✅ **Docker support** - Easy K8s deployment

**Deployment:**
```bash
# K8s pod with GPU
kubectl apply -f k8s/manifests/comfyui/
```

**Resources:**
- GitHub: https://github.com/comfyanonymous/ComfyUI
- Docker: `ghcr.io/ai-dock/comfyui:latest-cuda`
- API: https://github.com/comfyanonymous/ComfyUI/wiki/API

---

### 2. Character Consistency: IP-Adapter v2

**What is IP-Adapter?**
- Image prompt adapter for Stable Diffusion
- Maintains character appearance across generations
- Works with reference images (your existing luna.png, sakura.png)

**How it works:**
1. Load character reference image (luna.png)
2. IP-Adapter extracts face/style features
3. Apply to new generation → Same character, new expression

**Example Workflow:**
```python
# ComfyUI API call\nworkflow = {\n    \"reference_image\": \"luna.png\",\n    \"prompt\": \"1girl, happy expression, blushing, hand on cheek, anime style\",\n    \"negative_prompt\": \"sad, angry, distorted face\",\n    \"ip_adapter_strength\": 0.85,  # Character consistency\n    \"controlnet_pose\": \"neutral_pose.png\",\n    \"steps\": 20,\n    \"cfg_scale\": 7.5\n}\n```

**Resources:**
- GitHub: https://github.com/tencent-ailab/IP-Adapter
- Models: https://huggingface.co/h94/IP-Adapter
- ComfyUI node: Built-in

---

### 3. Pose Control: ControlNet OpenPose

**What is ControlNet?**
- Guides SD generation with structural hints (pose, edges, depth)
- Ensures consistent pose across expressions

**Workflow:**
1. Define base poses (idle, tilted head, covering face, etc.)
2. Store pose skeletons as ControlNet guides
3. Generate expressions while maintaining pose

**Example Poses for Waifu Animations:**
| Emotion | Pose | ControlNet Strength |\n|---------|------|--------------------|\n| Neutral | Standing straight | 0.7 |\n| Shy | Head down, hand covering mouth | 0.8 |\n| Flirty | Hip cocked, wink pose | 0.75 |\n| Excited | Arms raised, jumping | 0.85 |\n\n**Resources:**\n- GitHub: https://github.com/lllyasviel/ControlNet\n- Models: https://huggingface.co/lllyasviel/ControlNet-v1-1\n- Poses: https://github.com/fkunn1326/openpose-editor\n\n---\n\n## Model Selection\n\n### Recommended: SDXL-Turbo + AnimePastelDream LoRA\n\n**Base Model:** SDXL-Turbo (fast inference)\n**LoRA:** AnimePastelDream or similar anime-focused LoRA\n**ControlNet:** OpenPose v1.1\n**IP-Adapter:** IP-Adapter-Plus SDXL\n\n**Why this combo?**\n- Fast generation (2-4s per image)\n- Anime/waifu aesthetic\n- Character consistency\n- Pose control\n- 12GB VRAM fits perfectly\n\n**Resources:**\n- SDXL-Turbo: https://huggingface.co/stabilityai/sdxl-turbo\n- Anime LoRAs: https://civitai.com/tag/anime\n- Model size: ~8GB total (base + LoRA + ControlNet)\n\n---\n\n## Implementation Architecture\n\n```\n┌──────────────────────────────────────────────────────┐\n│ Frontend (React)                                     │\n│  ├─ Expression request (emotion + intensity)         │\n│  ├─ Image cache (recent expressions)                 │\n│  └─ Progressive loading (placeholder → full)         │\n└────────────────┬─────────────────────────────────────┘\n                 │ WebSocket: emotion change event\n                 ▼\n┌──────────────────────────────────────────────────────┐\n│ Backend (FastAPI)                                    │\n│  ├─ Expression cache (Redis)                         │\n│  │   - Key: waifu_id + emotion + intensity           │\n│  │   - TTL: 1 hour                                   │\n│  ├─ Generation queue (async)                         │\n│  │   - Priority: current emotion > prefetch          │\n│  └─ Prefetch logic                                   │\n│      - Generate common expressions at idle           │\n└────────────────┬─────────────────────────────────────┘\n                 │ HTTP: workflow + parameters\n                 ▼\n┌──────────────────────────────────────────────────────┐\n│ ComfyUI (K8s Pod with GPU)                           │\n│  ├─ SDXL-Turbo model (loaded in VRAM)                │\n│  ├─ IP-Adapter (character consistency)               │\n│  ├─ ControlNet (pose guidance)                       │\n│  ├─ LoRA (anime style)                               │\n│  └─ Workflow execution (2-4s per image)              │\n└──────────────────────────────────────────────────────┘\n```\n\n---\n\n## Generation Strategy\n\n### 1. Pre-Generation (Recommended)\n\n**Idea:** Generate all expression variations during deployment/idle time.\n\n**Expressions to pre-generate:**\n- 8 emotions × 3 intensities = 24 images per waifu\n- 2 waifus × 24 = 48 total images\n- Generation time: 48 × 3s = ~2.5 minutes\n- Storage: 48 × 2MB = ~96MB\n\n**Benefits:**\n- Zero latency during chat\n- Consistent quality\n- Predictable costs\n\n**Workflow:**\n```bash\n# Pre-generation script\npython scripts/generate_expressions.py \\\n  --waifu luna \\\n  --emotions neutral,happy,shy,flirty,excited,affectionate,playful,seductive \\\n  --intensities 0.3,0.6,0.9 \\\n  --output frontend/public/generated/\n```\n\n### 2. On-Demand Generation (Advanced)\n\n**For custom/rare expressions:**\n- User triggers unique emotion combination\n- Backend checks cache (Redis)\n- If not cached, queue generation\n- Show placeholder while generating\n- Update image when ready\n\n**Queue Priority:**\n1. Current emotion (high priority)\n2. Likely next emotions (medium)\n3. Rare combinations (low)\n\n---\n\n## Alternative: Live2D (Comparison)\n\n**Live2D Cubism:**\n- ✅ **Industry standard** - Used by major VTubers\n- ✅ **Smooth animations** - 60fps interpolation\n- ✅ **No GPU needed** - Runs on CPU/browser\n- ✅ **Rigged models** - Professional quality\n\n**Cons:**\n- 💰 **Expensive** - $42/mo Pro license + rigging costs ($100-1000)\n- ⚠️ **Complex pipeline** - Requires artist to rig models\n- ⚠️ **Less flexible** - Can't generate new characters easily\n\n**Use Case:** If budget allows and you want professional VTuber quality.\n\n**Resources:**\n- Website: https://www.live2d.com/\n- WebGL SDK: https://github.com/guansss/pixi-live2d-display\n- Rigging services: Fiverr, VGen\n\n**Verdict:** Live2D is higher quality but requires significant investment. Stable Diffusion offers more flexibility for dynamic content generation.\n\n---\n\n## Cost-Benefit Analysis\n\n### Option A: Pre-Generated SD Images (Recommended)\n\n**Costs:**\n- Development: 1-2 weeks\n- GPU time: ~3 minutes per character update\n- Storage: ~100MB per character\n- Ongoing: FREE (self-hosted)\n\n**Benefits:**\n- Zero latency\n- Full control\n- Easy to add new expressions\n- No recurring costs\n\n### Option B: On-Demand SD Generation\n\n**Costs:**\n- Development: 2-3 weeks (caching + queue)\n- GPU time: 2-4s per unique expression\n- Redis cache: Minimal\n- Ongoing: FREE (self-hosted)\n\n**Benefits:**\n- Infinite expression variations\n- Adaptive to user interactions\n- Storage efficient\n\n### Option C: Live2D Professional\n\n**Costs:**\n- License: $42/mo (Pro)\n- Rigging: $500-1000 per character (one-time)\n- Development: 3-4 weeks\n- Ongoing: $42/mo\n\n**Benefits:**\n- Professional VTuber quality\n- Smooth 60fps animations\n- Industry standard\n\n**Verdict:** For portfolio/demo → Option A (pre-gen). For production SaaS → Option B (on-demand). For premium product → Option C (Live2D).\n\n---\n\n## Implementation Plan\n\n### Phase 1: ComfyUI Deployment (Week 1)\n\n**Tasks:**\n- [ ] Deploy ComfyUI to K8s with GPU allocation\n- [ ] Install models (SDXL-Turbo, ControlNet, IP-Adapter)\n- [ ] Create workflow templates for each emotion\n- [ ] Test character consistency with luna.png\n- [ ] Benchmark generation speed\n\n**Deliverables:**\n- `k8s/manifests/comfyui/` - Deployment manifests\n- `workflows/` - ComfyUI workflow JSON files\n- `scripts/test_comfyui.py` - API test script\n\n### Phase 2: Expression Generation (Week 2)\n\n**Tasks:**\n- [ ] Define expression matrix (8 emotions × 3 intensities)\n- [ ] Create ControlNet pose guides\n- [ ] Generate all expressions for Luna and Sakura\n- [ ] Optimize prompts for consistency\n- [ ] Export to `frontend/public/generated/`\n\n**Deliverables:**\n- `scripts/generate_expressions.py` - Batch generation script\n- `frontend/public/generated/` - Expression images (48 files)\n- `docs/EXPRESSION_GUIDE.md` - Emotion mapping\n\n### Phase 3: Frontend Integration (Week 3)\n\n**Tasks:**\n- [ ] Update WaifuCanvas to load generated expressions\n- [ ] Implement expression switching logic\n- [ ] Add transition animations (fade/morph)\n- [ ] Test all emotion combinations\n- [ ] Optimize image loading (lazy load, progressive)\n\n**Deliverables:**\n- Updated `frontend/src/components/WaifuCanvas.js`\n- Expression transition system\n\n### Phase 4: Dynamic Generation (Optional, Week 4+)\n\n**Tasks:**\n- [ ] Build generation queue system\n- [ ] Add Redis caching\n- [ ] Implement prefetch logic\n- [ ] Add generation status UI\n- [ ] Load testing\n\n---\n\n## Performance Targets\n\n| Metric | Target | Current (Static) |\n|--------|--------|------------------|\n| **Expression latency** | <100ms (cached), <4s (generated) | Instant |\n| **Image quality** | 1024x2048 PNG | 1024x2048 PNG |\n| **Storage per waifu** | ~50MB (24 expressions) | ~2MB (1 image) |\n| **Generation throughput** | 15-20 images/min | N/A |\n| **VRAM usage** | <10GB | N/A |\n\n---\n\n## Resources\n\n**Tutorials:**\n- ComfyUI basics: https://www.youtube.com/watch?v=AbB33AxrcZo\n- IP-Adapter guide: https://civitai.com/articles/2792\n- ControlNet OpenPose: https://stable-diffusion-art.com/controlnet/\n\n**Model Libraries:**\n- Civitai (anime models): https://civitai.com/\n- HuggingFace: https://huggingface.co/models?pipeline_tag=text-to-image\n\n**Community:**\n- ComfyUI Discord: https://discord.gg/comfyui\n- r/StableDiffusion: https://reddit.com/r/StableDiffusion\n\n---\n\n## Next Steps\n\n1. **Deploy ComfyUI** - Set up on K8s with GPU\n2. **Test Character Consistency** - Validate IP-Adapter with luna.png\n3. **Generate Sample Set** - Create 8 expressions for Luna\n4. **User Feedback** - A/B test static vs. dynamic expressions\n\nShould I proceed with creating the ComfyUI K8s manifests and expression generation scripts?
